### PR TITLE
Scala 3

### PR DESCRIPTION
**Pull request by fdietze**
Everything works, but we found and reported a few compiler bugs on the way, which we found workarounds for:
- https://github.com/lampepfl/dotty/issues/14095 scala-dom-types implementation
- https://github.com/lampepfl/dotty/issues/14096 affects `render*` methods in `Outwatch.scala`
- https://github.com/lampepfl/dotty/issues/14169 affects rendering `js.UndefOr[VDomModifier]`
- https://github.com/lampepfl/dotty/issues/14171 found in the last test with Monix
- https://github.com/lampepfl/dotty/issues/15210 using `styleAttr <--` with Option[String]

To test this PR, here is a snapshot release: 
```scala
resolvers += ("Sonatype OSS Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots")
libraryDependencies += "io.github.outwatch"   %%% "outwatch" % "1.0.0-RC7+27-b57db7cf-SNAPSHOT"
```

### CODE REVIEW

**Comment by fdietze** [comment]
https://github.com/lampepfl/dotty/issues/14096
is affecting `render*` methods in `Outwatch.scala`

**Comment by fdietze** [comment]
Everything compiles now, but a single test is failing for all scala versions:
https://github.com/outwatch/outwatch/runs/4623399382?check_suite_focus=true#step:7:709

**Comment by fdietze** [comment]
Issue for rendering `js.UndefOr[VDomModifier]`:
https://github.com/lampepfl/dotty/issues/14169

**Comment by fdietze** [comment]
For the last test with Monix:
https://github.com/lampepfl/dotty/issues/14171

**Comment by cornerman** [comment]
Colibri in scala 3 will support the type classes `Source` and `Sink` as before. Though operations on `Observable` and `Observer` will require concrete types. Still can convert to them with `Observer.lift` and `Observable.lift`.

**Comment by busti** [comment]
@fdietze I have been working on a more generalized streaming abstraction over the past months in my fork of outwatch.  
But it would definitely benefit from scala 3, especially from `opaque` types and trait parameters.

File: build.sbt [hunk 1]
```diff
@@ -157,16 +164,16 @@ lazy val bench = project
...
-      ("jitpack" at "https://jitpack.io") ::
-      Nil,
+    resolvers += "jitpack" at "https://jitpack.io",
+    libraryDependencies ++= Seq(
+      "com.github.fdietze.bench" %%% "bench" % "d411db1",
+    ),
 
     libraryDependencies ++=
-      "com.github.fdietze.bench" %%% "bench" % "5ffab44" ::
+      "com.github.fdietze.bench" %%% "bench" % "d411db1" ::
```
Comment (cornerman) [inline build.sbt:173]: There are `libraryDependencies` twice :)

File: build.sbt [hunk 2]
```diff
@@ -196,7 +203,7 @@ lazy val docs = project
   .enablePlugins(MdocPlugin, DocusaurusPlugin)
   .settings(
     test/skip := true,
-    publish/skip := true,
+    publish/skip  := true,
```
Comment (cornerman) [inline build.sbt:206]: double space

File: outwatch/src/main/scala/outwatch/EmitterBuilder.scala [hunk 3]
```diff
@@ -38,14 +38,15 @@ import scala.concurrent.duration.FiniteDuration
...
+  @inline def forwardTo[F[_] : Sink, O2 >: O](sink: F[O2]): R
 
   // this method keeps the current Execution but actually, the caller must decide,
   // whether this really keeps the execution type or might be async. Therefore private.
   @inline private[outwatch] def transformWithExec[T](f: Observable[O] => Observable[T]): EmitterBuilderExecution[T, R, Exec]
   @inline private[outwatch] def transformSinkWithExec[T](f: Observer[T] => Observer[O]): EmitterBuilderExecution[T, R, Exec]
 
-  @inline def -->[F[_] : Sink](sink: F[_ >: O]): R = forwardTo(sink)
+  @inline def -->(sink: Observer[O]): R = forwardTo(sink)
+  @inline def -->[F[_] : Sink, O2 >: O](sink: F[O2], @annotation.nowarn dummy: Unit = ()): R = forwardTo(sink)
```
Comment (cornerman) [inline outwatch/src/main/scala/outwatch/EmitterBuilder.scala:50]: forwardTo should have the same overload.

File: outwatch/src/main/scala/outwatch/EmitterBuilder.scala [hunk 4]
```diff
@@ -300,8 +301,8 @@ object EmitterBuilder {
       }
     })
 
-  @noinline private def forwardToInTransform[F[_] : Sink, I, O, R: SubscriptionOwner](base: EmitterBuilder[I, R], transformF: Observable[I] => Observable[O], sink: F[_ >: O]): R = {
+  @noinline private def forwardToInTransform[F[_] : Sink, I, O, O2 >: O, R: SubscriptionOwner](base: EmitterBuilder[I, R], transformF: Observable[I] => Observable[O], sink: F[O2]): R = {
     val connectable = Observer.lift(sink).redirect(transformF)
-    SubscriptionOwner[R].own(base.forwardTo(connectable.value))(connectable.connect)
+    SubscriptionOwner[R].own(base.forwardTo(connectable.value))(() => connectable.connect())
```
Comment (cornerman) [inline outwatch/src/main/scala/outwatch/EmitterBuilder.scala:306]: Is this needed?

File: snabbdom/src/main/scala/snabbdom/Snabbdom.scala [hunk 5]
```diff
@@ -89,13 +88,13 @@ object thunk {
...
     } {
-      prepatch(fn, shouldRender, oldProxy, thunk)
+      thunkPrepatch(fn, shouldRender, oldProxy, thunk)
     }
 
     thunk.data.foreach(_.hook.foreach(_.prepatch.foreach(_ (oldProxy, thunk))))
   }
 
-  @inline private def prepatch(fn: () => VNodeProxy, shouldRender: Boolean, oldProxy: VNodeProxy, thunk: VNodeProxy): Unit = {
+  @inline private def thunkPrepatch(fn: () => VNodeProxy, shouldRender: Boolean, oldProxy: VNodeProxy, thunk: VNodeProxy): Unit = {
```
Comment (cornerman) [inline snabbdom/src/main/scala/snabbdom/Snabbdom.scala:97]: Why did we need to rename this?

File: tests/src/test/scala/outwatch/ScenarioTestSpec.scala [hunk 6]
```diff
@@ -235,31 +237,34 @@ class ScenarioTestSpec extends JSDomAsyncSpec {
       clickStream <- Handler.createF[IO, MouseEvent]
       keyStream <- Handler.createF[IO, KeyboardEvent]
 
-      buttonDisabled = textFieldStream
+
+    } yield {
```
Comment (cornerman) [inline tests/src/test/scala/outwatch/ScenarioTestSpec.scala:241]: Why did we rewrite this? :)

File: snabbdom/src/main/scala/snabbdom/Snabbdom.scala (hunk 5 above)
Comment (fdietze) [reply snabbdom/src/main/scala/snabbdom/Snabbdom.scala:97]: there was a name collision, I think.

File: tests/src/test/scala/outwatch/ScenarioTestSpec.scala (hunk 6 above)
Comment (fdietze) [reply tests/src/test/scala/outwatch/ScenarioTestSpec.scala:241]: because of this compiler bug:
https://github.com/lampepfl/dotty/issues/14171

**Review by busti — APPROVED** [review]

**Comment by busti** [comment]
Looks like it fails due to some faulty jitpack dependencies.  
I propose that we replace these with git submodules and `ProjectRef`

File: outwatch/src/main/scala/outwatch/helpers/NativeHelpers.scala [hunk 7]
```diff
@@ -33,8 +33,8 @@ private[outwatch] object NativeHelpers {
   }
 
   @noinline def prependSeq[T](source: js.Array[T], other: collection.Seq[T]): js.Array[T] = if (other.isEmpty) source else other match {
-    case wrappedOther:js.WrappedArray[T] =>
-      if (source.isEmpty) wrappedOther else (wrappedOther: js.Array[T]).concat(source)
+    case wrappedOther:js.WrappedArray[_] =>
```
Comment (cornerman) [inline outwatch/src/main/scala/outwatch/helpers/NativeHelpers.scala:36]: Why was this change needed?

Comment (fdietze) [reply outwatch/src/main/scala/outwatch/helpers/NativeHelpers.scala:36]: yes, else the error is:

```scala
[error] [E1] outwatch/src/main/scala/outwatch/helpers/NativeHelpers.scala
[error]      the type test for scalajs.js.WrappedArray[T] cannot be checked at runtime
[error]      L36:     case wrappedOther:js.WrappedArray[T] =>
[error]                    ^
```

Comment (cornerman) [reply outwatch/src/main/scala/outwatch/helpers/NativeHelpers.scala:36]: Ah yeah. Does it work with the unchecked annotation?
```case wrapperOther:js.WrapperArray[T@unchecked]```
